lib/pbio/drv/usb/usb.c

```diff
@@ -0,0 +1,227 @@
+// SPDX-License-Identifier: MIT
+// Copyright (c) 2025 The Pybricks Authors
+
+#include <pbdrv/config.h>
+
+#if PBDRV_CONFIG_USB
+
+#include <errno.h>
+#include <signal.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+
+#include "usb.h"
+
+#include <pbdrv/config.h>
+#include <pbdrv/usb.h>
+
+#include <pbio/error.h>
+#include <pbio/os.h>
+#include <pbio/protocol.h>
+
+#include <lwrb/lwrb.h>
+
+/**
+ * Host is subscribed to our outgoing event messages.
+ */
+static bool pbdrv_usb_events_subscribed;
+
+bool pbdrv_usb_connection_is_active(void) {
+    return pbdrv_usb_events_subscribed && pbdrv_usb_is_ready();
+}
+
+/**
+ * Pybricks system command handler.
+ */
+static pbdrv_usb_receive_handler_t pbdrv_usb_receive_handler;
+
+void pbdrv_usb_set_receive_handler(pbdrv_usb_receive_handler_t handler) {
+    pbdrv_usb_receive_handler = handler;
+}
+
+/**
+ * Buffer scheduled status.
+ */
+static uint8_t status_data[PBIO_PYBRICKS_EVENT_STATUS_REPORT_SIZE];
+static bool status_data_pending;
+
+void pbdrv_usb_schedule_status_update(const uint8_t *status_msg) {
+    // Ignore if message identical to last.
+    if (!memcmp(status_data, status_msg, sizeof(status_data))) {
+        return;
+    }
+
+    // Schedule to send whenever the Bluetooth process gets round to it.
+    memcpy(status_data, status_msg, sizeof(status_data));
+    status_data_pending = true;
+    pbio_os_request_poll();
+}
+
+/**
+ * Buffer for scheduled stdout.
+ */
+static lwrb_t stdout_ring_buf;
+
+pbio_error_t pbdrv_usb_stdout_tx(const uint8_t *data, uint32_t *size) {
+
+    if (!pbdrv_usb_connection_is_active()) {
+        return PBIO_ERROR_INVALID_OP;
+    }
+
+    // Buffer data to send it more efficiently even if the caller is only
+    // writing one byte at a time.
+    if ((*size = lwrb_write(&stdout_ring_buf, data, *size)) == 0) {
+        return PBIO_ERROR_AGAIN;
+    }
+
+    // Poke the process to start tx soon-ish. This way, we can accumulate
+    // data bytes before actually transmitting.
+    pbio_os_request_poll();
+
+    return PBIO_SUCCESS;
+}
+
+uint32_t pbdrv_usb_stdout_tx_available(void) {
+    if (!pbdrv_usb_connection_is_active()) {
+        return UINT32_MAX;
+    }
+    return lwrb_get_free(&stdout_ring_buf);
+}
+
+bool pbdrv_usb_stdout_tx_is_idle(void) {
+    if (!pbdrv_usb_connection_is_active()) {
+        return true;
+    }
+    return lwrb_get_full(&stdout_ring_buf) == 0;
+}
+
+static bool respond_soon;
+static pbio_pybricks_error_t respond_result;
+
+/**
+ * Non-blocking poll handler to process pending incoming messages.
+ */
+static void pbdrv_usb_handle_data_in(void) {
+
+    // Ignore incoming data if we haven't sent our previous response yet.
+    if (respond_soon) {
+        return;
+    }
+
+    static uint8_t data_in[PBDRV_CONFIG_USB_MAX_PACKET_SIZE];
```

Comment:
Would benefit from a comment explaining why this is static and why we need to make a copy of it.